prov/util: Add av_entry_size field to struct util_av_attr

dipti-kothari opened the pull request:
To allow custom av entry size, add an additional field
i.e. av_entry_size to struct util_av_attr

Signed-off-by: Dipti Kothari <dkothar@amazon.com>

shefty commented:
Nothing is added that makes use of this, so it's hard to evaluate if this is the right change that's needed.  The AV entry is already larger than the address, as struct util_av_entry is also there.  It may be possible to simply add a new field to that structure.

Even if we decide to go with the proposed changes, the name av_entry_size is misleading.  The caller only needs to specify how many extra bytes of data need to be stored with each address.  The actual av entry size is something internal (based on the address length + struct util_av_entry + possible some additional data).

Related to either option above, the resulting struct util_av_entry that is retrieved from the buffer pool ends up with a misidentified field.  Any extra data ends up being listed as part of the 'addr' field of that structure.

dipti-kothari commented:
> Nothing is added that makes use of this, so it's hard to evaluate if this is the right change that's needed.

I have another [PR 6022](https://github.com/ofiwg/libfabric/pull/6022) which adds resizing logic for address vector in efa provider. PR 6022 has the commit from this PR as well.
> The AV entry is already larger than the address, as struct util_av_entry is also there. It may be possible to simply add a new field to that structure.

 EFA has some additional fields  in struct efa_av_entry other than the address. 

`struct efa_av_entry {
        uint8_t                 ep_addr[EFA_EP_ADDR_LEN];
        fi_addr_t               rdm_addr;
        fi_addr_t               shm_rdm_addr;
        bool                    local_mapping;
};`
I can add a new field to util_av_entry instead of util_av_attr. That works too.
> 
> Even if we decide to go with the proposed changes, the name av_entry_size is misleading. The caller only needs to specify how many extra bytes of data need to be stored with each address. The actual av entry size is something internal (based on the address length + struct util_av_entry + possible some additional data).

The additional data field can be renamed to context_len instead of av_entry_size
> 
> Related to either option above, the resulting struct util_av_entry that is retrieved from the buffer pool ends up with a misidentified field. Any extra data ends up being listed as part of the 'addr' field of that structure.

Should we rename 'addr' to something more generic in util_av_entry to allow additional fields?

dipti-kothari commented on include/ofi_util.h:
```diff
@@ -707,6 +707,7 @@ struct util_av {
 
 struct util_av_attr {
 	size_t	addrlen;
+	size_t  av_entry_size;
```
Added a comment, Thanks

shefty commented:
Yes, I would rename addr to something more generic (e.g. data), since both addressing and context information will be stored there.  Renaming av_entry_size to context_len would make it clearer that it is only specifying the extra context size that's needed.

I would add a comment above the renamed addr field to describe that the address comes first, followed by extra provider context information.  The code will also need to ensure that extra context data is aligned on a reasonable boundary, such as 8-bytes.  This could be done by asserting that the address length is a multiple of 8-bytes, or forcing the offset.

dipti-kothari commented:
> Yes, I would rename addr to something more generic (e.g. data), since both addressing and context information will be stored there. Renaming av_entry_size to context_len would make it clearer that it is only specifying the extra context size that's needed.

@shefty, I have update the PR, renamed 'addr' to 'data' and 'av_entry_size' to 'context_len'. Since we have addrlen in struct util_av_attr, I have added context_len field also in struct util_av_attr.

dipti-kothari commented:
> I would add a comment above the renamed addr field to describe that the address comes first, followed by extra provider context information. The code will also need to ensure that extra context data is aligned on a reasonable boundary, such as 8-bytes. This could be done by asserting that the address length is a multiple of 8-bytes, or forcing the offset.

I would probably have to add the assertion in ofi_av_insert_addr instead of util_av_init. I cannot increase the addrlen for shm provider.

zhngaj commented on prov/shm/src/smr_av.c:
```diff
@@ -220,7 +220,8 @@ int smr_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
 		return -FI_ENOMEM;
 
-	util_attr.addrlen = NAME_MAX;
+	util_attr.addrlen = NAME_MAX + 1;
```
Sorry, did not realize that it's merged already.
This change looks like a hack, just to make the addrlen is multiple of 8 bytes. Do we really want to do this or is there something I'm missing?

shefty replied:
An assert was added to the util av code to check that the addrlen is a multiple of 8-bytes.  Though it likely only needs to be a multiple of sizeof(void *).  This was the easy change.

A more complex change would be to remove the addrlen size restriction, calculate an alignment offset for context_len, and align context that way.

Note that NAME_MAX excludes the null terminating character.  So adding 1 here actually makes some sense if the address includes it.  I can't easily figure out the intent of the shm code here.  In some cases, it uses NAME_MAX directly, in other cases it does NAME_MAX - 1.

dipti-kothari replied:
Yeah, I can instead add an additional field for 'addr_offset', that seems like a better solution

```
diff --git a/include/ofi_util.h b/include/ofi_util.h
index 0f57a851d..2870336e2 100644
--- a/include/ofi_util.h
+++ b/include/ofi_util.h
@@ -707,6 +708,7 @@ struct util_av {
        uint64_t                flags;
        size_t                  count;
        size_t                  addrlen;
+        size_t                  addr_offset;
        struct dlist_entry      ep_list;
        fastlock_t              ep_list_lock;
 };

diff --git a/prov/util/src/util_av.c b/prov/util/src/util_av.c
index eefd5cf0e..cd29fe9cf 100644
--- a/prov/util/src/util_av.c
+++ b/prov/util/src/util_av.c
@@ -269,7 +269,7 @@ int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
 {
        struct util_av_entry *entry = NULL;

-       HASH_FIND(hh, av->hash, addr, av->addrlen, entry);
+       HASH_FIND(hh, av->hash, addr, av->addrlen + av->addr_offset, entry);
        if (entry) {
                if (fi_addr)
                        *fi_addr = ofi_buf_index(entry);
@@ -281,9 +281,9 @@ int ofi_av_insert_addr(struct util_av *av, const void *addr, fi_addr_t *fi_addr)
                        return -FI_ENOMEM;
                if (fi_addr)
                        *fi_addr = ofi_buf_index(entry);
-               memcpy(entry->data, addr, av->addrlen);
+               memcpy(entry->data, addr, av->addrlen + av->addr_offset);
                ofi_atomic_initialize32(&entry->use_cnt, 1);
-               HASH_ADD(hh, av->hash, data, av->addrlen, entry);
+               HASH_ADD(hh, av->hash, data, av->addrlen + av->addr_offset, entry);
        }
        return 0;
 }

@@ -423,10 +423,14 @@ static int util_av_init(struct util_av *av, const struct fi_av_attr *attr,
 {
        int ret = 0;
        size_t max_count;
+       size_t offset;

-       assert(util_attr->addrlen % 8 == 0);
+       /* offset calculated on a 8-byte boundary */
+       offset = util_attr->addrlen % 8;
+       if (offset != 0)`
+               offset = 8 - offset;
        struct ofi_bufpool_attr pool_attr = {
-               .size           = util_attr->addrlen +
+              .size           = util_attr->addrlen + offset +
                                  util_attr->context_len +
                                  sizeof(struct util_av_entry),
                .alignment      = 16,
@@ -449,6 +453,7 @@ static int util_av_init(struct util_av *av, const struct fi_av_attr *attr,
        FI_INFO(av->prov, FI_LOG_AV, "AV size %zu\n", av->count);
        av->addrlen = util_attr->addrlen;
+       av->addr_offset = offset;
```
Please let me know what you think about it

shefty replied:
The address should be aligned okay, since it's at the top of the data field.  The problem is that the context may be misaligned.  If the provider is expecting to cast &data[addrlen] to a structure, we need that aligned.  Switching to &data[context_offset] works.  We need to adjust the allocated data size based on extra alignment space.
